### PR TITLE
[ParseableInterfaces] Handle lazy vars

### PR DESCRIPTION
**Pull request by harlanhaskins**
For lazy vars, we need to make public the top-level entry point, and the fact that the lazy storage contributes to the layout of a type (if it’s fixed layout, or if the type is not resilient.) We also shouldn’t ever print `lazy` in a parseable interface.

Since we need to parse these identifiers, give them a new name, `$__lazy_storage_{propname}`, which is parseable only in parseable interfaces so it doesn’t conflict with other properties.

### CODE REVIEW

File: lib/Sema/CodeSynthesis.cpp [hunk 1]
```diff
@@ -1346,8 +1346,9 @@ void swift::completeLazyVarImplementation(VarDecl *VD) {
   assert(!VD->isStatic() && "Static vars are already lazy on their own");
 
   // Create the storage property as an optional of VD's type.
-  SmallString<64> NameBuf = VD->getName().str();
-  NameBuf += ".storage";
+  SmallString<64> NameBuf;
+  NameBuf += "$__lazy_storage_";
```
Comment (jrose-apple) [inline lib/Sema/CodeSynthesis.cpp]: Nitpick: maybe separate the prefix from the name even harder? ObjC does `_$_` between sections, so `__lazy_storage_$_foo` or `$__lazy_storage_$_foo`. The former also makes it even less likely to conflict with a debugger variable, which probably only has the `$` at the beginning.

File: test/ParseableInterface/lazy-vars.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+// RUN: %FileCheck %s < %t/Test.swiftinterface --check-prefix CHECK --check-prefix NONRESILIENT
+// RUN: %target-swift-frontend -build-module-from-parseable-interface %t/Test.swiftinterface -o %t/Test.swiftmodule
+// RUN: %target-swift-frontend -emit-module -o /dev/null -merge-modules -emit-parseable-module-interface-path - %t/Test.swiftmodule -module-name Test | %FileCheck %s --check-prefix CHECK --check-prefix NONRESILIENT
+
+// RUN: %target-swift-frontend -typecheck -module-name TestResilient -emit-parseable-module-interface-path %t/TestResilient.swiftinterface -enable-resilience %s
+// RUN: %FileCheck %s < %t/TestResilient.swiftinterface --check-prefix CHECK --check-prefix RESILIENT
+
+// RUN: %target-swift-frontend -build-module-from-parseable-interface %t/TestResilient.swiftinterface -o %t/TestResilient.swiftmodule
+// RUN: %target-swift-frontend -emit-module -o /dev/null -merge-modules -emit-parseable-module-interface-path - %t/TestResilient.swiftmodule -module-name TestResilient | %FileCheck %s --check-prefix CHECK --check-prefix RESILIENT --implicit-check-not NONRESILIENT
```
Comment (jrose-apple) [inline test/ParseableInterface/lazy-vars.swift]: `--implicit-check-not` is a single pattern that's inserted between every positive check, not a prefix of things to omit. I think it's okay to just not have this.

File: lib/Sema/CodeSynthesis.cpp (hunk 1 above)
Comment (harlanhaskins) [reply lib/Sema/CodeSynthesis.cpp]: I'm hesitant to write it without the `$` at the beginning, as that makes it a valid identifier. Do you think the potential name conflict matters there?

Comment (jrose-apple) [reply lib/Sema/CodeSynthesis.cpp]: Ah, I forgot we allowed `$` in non-initial positions. Okay.

**Review by jrose-apple — APPROVED** [review]

**Pull request merged**